### DISCUSSION TITLE
QST: What polygon ring nesting standard??

### DISCUSSION THREAD

**Discussion by thomcom** [geopandas/geopandas#2727] [Q&A]
- [x] I have searched the [geopandas] tag on [StackOverflow](https://stackoverflow.com/questions/tagged/geopandas) and [GIS StackExchange](https://gis.stackexchange.com/questions/tagged/geopandas) for similar questions.

- [x] I have asked my usage related question on [StackOverflow](https://stackoverflow.com) or [GIS StackExhange](https://gis.stackexchange.com).

---

Hey all. We're having an interesting conversation about polygon ring winding over at `cuspatial` and I wanted to touch base here. In particular we're talking about [this post](https://gis.stackexchange.com/questions/119150/order-of-polygon-vertices-in-general-gis-clockwise-or-counterclockwise) that describes a Simple Feature format and an ESRI Shapefile format for polygons. The punchline is that:

Simple Feature format uses a counter-clockwise outer ring and clockwise inner rings. The first ring must be the exterior polygon, subsequent rings must be inner polygons. Island-in-a-lake-in-an-island patterns must be specified with MultiPolygon.

ESRI Shapefile format uses `n` clockwise outer rings and `m` counter-clockwise inner rings. The first ring must be exterior, but subsequent rings can also be exterior. This makes every ESRI Polygon implicitly a MultiPolygon. island-in-a-lake-in-an-island is possible with a single Polygon.

Have ya'll specified exactly which rule you follow? I'd like to copy your wisdom as always. :) I'd also like to write tests over at `cuspatial` that validate that a specific format has been chosen and enforced.

**Comment by jorisvandenbossche** [comment]
On this front, GeoPandas actually has no "opinion" and is flexible to work with whatever, as that is what the underlying GEOS library does. 

GEOS doesn't care about the orientation or rings and can work with either orientation standards (or even something in between, like both outer and inner rings having the same orientation). However, when it produces new geometries itself during operations, it will typically (and a bit unfortunately) create geometries using the opposite of the Simple Feature standard (so basically ESRI standard, expect that GEOS polygons only have a single exterior ring, like in the Simple Features standard). 

And because GEOS only deals with geometries on the cartesian plane, this is actually OK for GEOS to not care (it's only on the sphere that you need a specific orientation rule to be able to interpret a ring). Of course, also when dealing with cartesian coordinates, you could write algorithms that expect a certain orientation to work properly (but so GEOS doesn't do that AFAIK).

**Comment by jorisvandenbossche** [comment]
There was recently some conversation about this on the GEOS matrix/element.io chat channel: https://matrix.to/#/!xiBlfkzpldcIeZqgQI:osgeo.org/$0_WKaXLAm4rq3mwc6WAsGGmC3MHED0HPhaGVzYltd1g?via=libera.chat&via=matrix.org&via=osgeo.org

Quoting one of the answers of Martin Davis:

> As others have said, GEOS doesn't care about polygon ring orientation - it checks orientation before carrying out computations that depend on it, and can handle any orientation.
When it creates polygonal output, it does enforce an orientation - CW for exterior rings, CCW for holes.
This is opposite to the **current** OGC standard, since when JTS/GEOS was initially developed the OGC SFA spec didn't specify ring orientation. Unfortunately we guessed wrong... 😞
Some systems are sensitive to ring orientation - you need to know what you're dealing with

**Comment by thomcom** [comment]
Thanks for the quick response! Judging from your answer and opinions internally, I'm going to leave it open and undefined in my mind despite what the "accepted answer" in [the gis.stackoverflow.com post](https://gis.stackexchange.com/questions/119150/order-of-polygon-vertices-in-general-gis-clockwise-or-counterclockwise) claimed. An undefinable standard, albeit a small one. :) I call those towers-of-babel.

**Comment by jorisvandenbossche** [comment]
Note that the "accepted" answer (not the first one) on that SO question is outdated (it has an updated note, but the starting paragraph about saying there is no standard in OGC is no longer correct). 

I think most software that deals with geographic coordinates (so on a sphere) enforces the OGC standard of counter-clockwise exterior ring + clockwise interior rings (eg S2, BigQuery: https://cloud.google.com/bigquery/docs/geospatial-data#polygon_orientation)
